### ISSUE TITLE
Examples are panicking when running in virtualbox

### ISSUE THREAD

**Issue by itgenie98** [bevyengine/bevy#413]
Virtualbox 6.1.12
Guest OS: Arch Linux, Kernel 5.8.5
GPU: VMware SVGA II Adapter (VMSVGA)

Steps to reproduce:
1. use a virtualbox vm with the VMSVGA graphics adapter
2. install rustup
3. clone bevy repository (for the examples)
4. run `cargo run --example breakout`

Expected behavior:
Running the bevy breakout example

Actual behavior:
Cargo / Rust crashes with following errormessage: (with `RUST_BACKTRACE=1`)
```
     Running `target/debug/examples/breakout`
thread 'main' panicked at 'Unable to find a GPU! Make sure you have installed required drivers!', crates/bevy_wgpu/src/wgpu_renderer.rs:28:14
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1076
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1537
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:217
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:526
  11: rust_begin_unwind
             at src/libstd/panicking.rs:437
  12: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
  13: core::option::expect_failed
             at src/libcore/option.rs:1261
  14: core::option::Option<T>::expect
             at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libcore/option.rs:344
  15: bevy_wgpu::wgpu_renderer::WgpuRenderer::new::{{closure}}
             at crates/bevy_wgpu/src/wgpu_renderer.rs:22
  16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libcore/future/mod.rs:78
  17: pollster::block_on
             at /home/itgenie98/.cargo/registry/src/github.com-1ecc6299db9ec823/pollster-0.2.0/src/lib.rs:64
  18: bevy_wgpu::wgpu_render_system
             at crates/bevy_wgpu/src/lib.rs:35
  19: <bevy_wgpu::WgpuPlugin as bevy_app::plugin::Plugin>::build
             at crates/bevy_wgpu/src/lib.rs:22
  20: bevy_app::app_builder::AppBuilder::add_plugin
             at ./crates/bevy_app/src/app_builder.rs:235
  21: <bevy_app::app_builder::AppBuilder as bevy::add_default_plugins::AddDefaultPlugins>::add_default_plugins
             at src/add_default_plugins.rs:33
  22: breakout::main
             at examples/game/breakout.rs:9
  23: std::rt::lang_start::{{closure}}
             at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/rt.rs:67
  24: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
  25: std::panicking::try::do_call
             at src/libstd/panicking.rs:348
  26: std::panicking::try
             at src/libstd/panicking.rs:325
  27: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  28: std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  29: std::rt::lang_start
             at /rustc/04488afe34512aa4c33566eb16d8c912a3ae04f9/src/libstd/rt.rs:67
  30: main
  31: __libc_start_main
  32: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by M4SSD35TRUCT10N** [comment]
Seems related to #402. Doesn't work on x86_64 Manjaro XFCE 20.0.3 nor on Raspberry Pi 4B+ with 8GB RAM either. Although it compiles without any problems.

**Comment by jamadazi** [comment]
Are you sure the virtual machine "gpu" and its driver support Vulkan? I doubt that's the case, because IIRC vulkan support in virtual machines is still very experimental if at all existent.

Bevy does not support an OpenGL or GLES backend yet, or any sort of software rendering, so you cannot run it if you don't have a GPU and driver that support Vulkan (or dx12 on windows, or metal on mac).

I wouldn't expect bevy to work in a virtual machine at all, at least not yet.

**Comment by M4SSD35TRUCT10N** [comment]
> Are you sure the virtual machine "gpu" and its driver support Vulkan? I doubt that's the case, because IIRC vulkan support in virtual machines is still very experimental if at all existent.
> 
> Your error seems to indicate that this is the case ("cannot find gpu!"); your system does not provide the needed GPU APIs.
> 
> Bevy does not support an OpenGL or GLES backend yet, or any sort of software rendering, so you cannot run it if you don't have a GPU and driver that support Vulkan (or dx12 on windows, or metal on mac).
> 
> I wouldn't expect bevy to work in a virtual machine at all, at least not yet.

I do have vulkan-intel driver installed on my Laptop. Currently checking out if it can be or is already loaded. On Raspberry Pi 4B+ it's certainly not available and thus understandable that it doesn't work. But anyway @jamadazi thank you for the hint.

**Comment by mkesper** [comment]
Is is possible to install mesa-vulkan-drivers? That's what it's called in Debian. Performance will be bad but at least it does run.

**Comment by CleanCut** [comment]
@mkesper I've got `mesa-vulkan-drivers` installed in my Ubuntu 20.04 VM, and it fails with the same "Unable to find a GPU! ..." error. 😢

**Comment by mkesper** [comment]
Sorry, I've no idea then, that was just what helped me on baremetal.

**Comment by code-monad** [comment]
# TL;DR
It's not bevy's fault, this is due to the SVGA driver does not support vulkan right now.
## Methods to avoid this issue(currently):
- Painless: Get a GPU passthrough to your guest & install the proper driver.
- Software layer: Install a software Vulkan rasterizer, more specifically, lavapipe or the [mesa main branch](https://gitlab.freedesktop.org/mesa/mesa/-/tree/main) which have merged the lavapipe features. But you'll meet **performance issue**(quite obviously), and make sure that you're walking on the edge
- New backend or layer: this might be a hardcore job since we are using `wgpu`,  which may still having issue about running over [OpenGLES on Linux](https://github.com/gfx-rs/wgpu/issues/1551), you may implement a software render backend, and it's definetly hard._\*sigh\*_
-----

Although this is a thread from September, but it is still open, and I've ran into the same issue, tried to figure it out with some workaround.

The env I've tried:
1. WSL2 on Win10/Win11 with a nVidia GPU on host
> - All presentation is through  X410(A XServer on Windows)
> - Tried to install the experimental CUDA driver that supports WSL, but with no luck

2. Archlinux on a VMWare host
> - Same as previous, no WM&DE, just use a XServer on my host machine for more flexible and my personal development habit.
> - Key packages I've installed: `mesa`, `vulkan-swrast`, and with an optional `vulkan-mesa-layers`

And here are several ways about how to make bevy(or more _generically_, **vulkan backend**) running under a VM-like environment without hacking the source code of bevy:

# 1. Try to get a external gpu and passthrough it to you guest machine.
  This is the most efficient and painless way to avoid the none gpu issue. You will fell like you are getting the power of the computed units -- everything even cuda works fine.
  For about how to configure PCIE passthrough , here are some links about it:
- [Virtual Box's manual about PCIE passthrogh](https://docs.oracle.com/en/virtualization/virtualbox/6.0/admin/pcipassthrough.html), and it's **only** working on a **Linux host**
- [VMware(ESXi) guide](https://kb.vmware.com/s/article/1010789), note that VMWare Workstation Pro does not support that.
- [PCI passthrough via OVMF](https://wiki.archlinux.org/title/PCI_passthrough_via_OVMF)
- Or, if you're lucky & rich enough:  [Installing and configuring the NVIDIA VIB on ESXi](https://kb.vmware.com/s/article/2033434)

And do make sure you installed the proper driver, check if the dkms module is loaded, `vkcube` works fine and runnign in a high refresh rate! It's cellarbration time!  You can even connect the GPU output ports to external display monitors

But the problem is, GPU is really expensive right now and you need a external GPU with a second PCIE slot to do that, which may only in full-sized boards, no luck for ITX builds and low-end products.

# 2. Configuring a software Vulkan rasterizer
So we are here, we do not own any other GPUs, or we can't do that, or you just don't want to, whatever the reason is, we need to fix it and get bevy, and other Vulkan applications working on a VM guest.

If you are in Ubuntu/WSL/WSL2, you could try compile the mesa source yourself and install it, or you could just use this PPA: https://launchpad.net/~kisak/+archive/ubuntu/kisak-mesa

If you are in Arch, just install `vulkan-swrast` after you installing the `xf86-video-vmware`

The FPS is about half of the hardware solution, you can fell you are back to the 80s while running the `breakout` exmaple. But hey, anyway we successfully make it running without complains!
\*Forget about the single line says ```WARNING: lavapipe is not a conformant vulkan implementation, testing use only.``` and do hope yourself lucky all the time.\*

**Comment by mockersf** [comment]
Bevy has a few examples running in CI using `lava pipe`, you can check the job here: https://github.com/bevyengine/bevy/blob/073f381c9ee2236ab0ce92a075422ce302f0468c/.github/workflows/ci.yml#L216-L259

**Comment by makspll** [comment]
If you stumbled upon this later, I believe the current version of the code linked above lives in:

https://github.com/bevyengine/bevy/blame/236091adce9c7a21684e5ed385fc180160bd047a/.github/workflows/example-run.yml#L85C14-L85C14
